pkg/stash/with_locker.go

```diff
@@ -0,0 +1,133 @@
+package stash
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/gomods/athens/pkg/errors"
+	"github.com/gomods/athens/pkg/observ"
+	"github.com/gomods/athens/pkg/storage"
+)
+
+const (
+	defaultGetLockTimeout = 5 * time.Minute
+	defaultPingInterval   = 10 * time.Second
+)
+
+type locker interface {
+	// lock obtains a lock to be used for single flight.
+	// The lock will be released when ctx is canceled. When the lock is released whether intentionally or not,
+	// releaseErrs will receive an error. If that error is nil, the lock was successfully released as a response to ctx.
+	// Returns:
+	//	 releaseErrs:	A channel that emits an error when the lock is released.
+	//   err: 			Any error creating the lock. If err is non-nil, no lock is created.
+	lock(ctx context.Context, name string) (releaseErrs <-chan error, err error)
+}
+
+var ErrUnexpectedRelease = fmt.Errorf("lock was unexpectedly released")
+
+func withLocker(l locker, checker storage.Checker) Wrapper {
+	return func(s Stasher) Stasher {
+		return &lockerLock{
+			locker:  l,
+			stasher: s,
+			checker: checker,
+		}
+	}
+}
+
+type lockerLock struct {
+	locker  locker
+	stasher Stasher
+	checker storage.Checker
+}
+
+func (s *lockerLock) Stash(ctx context.Context, mod, ver string) (string, error) {
+	const op errors.Op = "lockerLock.Stash"
+	ctx, span := observ.StartSpan(ctx, op.String())
+	defer span.End()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	mv := fmt.Sprintf("%s@%s", mod, ver)
+	releaseErrs, err := s.locker.lock(ctx, mv)
```

Comment:
This interaction of this function API seems a bit weird. We ask it to lock for us, and gives us an immediate error plus an asynchronous error that we have to wait on in a goroutine. 

Why not instead of this, we make `s.locker.lock` take a context (which it already does) but instead return a new context. 

This will mimic a lot of std lib behavior like `context.WithTimeout` and `context.WithCancel`. 

This way, users of this API won't be confused as to how they are supposed to use this function. They can just do this: 

```
ctx, err = s.locker.lock(ctx, mv)
if err != nil { ... }
// pass new ctx that will handle lock cancellation...
```